### PR TITLE
Fix leaks2

### PR DESCRIPTION
**Pull request by Therzok**
Fixed leaks surrounding editing files with projections and ~a leak of a TextBufferUndoManager for each document that was opened.~ (filed [upstream](https://github.com/Microsoft/vs-editor-api/issues/8))

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 1]
```diff
@@ -171,9 +171,11 @@ Task<SolutionInfo> CreateSolutionInfo (MonoDevelop.Projects.Solution solution, C
 			return Task.Run (async delegate {
 				var projects = new ConcurrentBag<ProjectInfo> ();
 				var mdProjects = solution.GetAllProjects ();
-				foreach (var p in projectionList)
-					p.Dispose ();
-				projectionList = projectionList.Clear ();
+				lock (projectionListUpdateLock) {
+					foreach (var p in projectionList)
+						p.Dispose ();
```
Comment (DavidKarlas) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Can you make copy of list and dispose that list outside `lock`, just to be safe...

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Done.

**Comment by Therzok** [comment]
Seems like the TextBufferUndoManager leak is in upstream.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs [hunk 2]
```diff
@@ -171,9 +171,14 @@ Task<SolutionInfo> CreateSolutionInfo (MonoDevelop.Projects.Solution solution, C
 			return Task.Run (async delegate {
 				var projects = new ConcurrentBag<ProjectInfo> ();
 				var mdProjects = solution.GetAllProjects ();
+				ImmutableList<ProjectionEntry> toDispose;
+				lock (projectionListUpdateLock) {
+					toDispose = projectionList;
```
Comment (KirillOsenkov) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Do you ever do anything with toDispose?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.TypeSystem/MonoDevelopWorkspace.cs]: Forgot to rename! good catch

**Pull request merged**